### ISSUE TITLE
[BUG] pip broke painless source code dependencies for Cython, there needs to be a way to do this that works in practice

### ISSUE THREAD

**Issue by ell1e** [cython/cython#5701]
### Describe the bug

Pip added build isolation, and sadly this broke painless source code dependencies for Cython for cross-project `.pxd` import because now every such dependency needs to be fully built (which if it's Cython can take forever) *again* for each install due to build isolation. Pip has also decided that build isolation is always force-enabled for all packages now, so this makes it impossible to make a Cython package depend on another Cython package for `.pxd` import without causing the user potentially really extreme and untenable build times. There needs to be a way to do this that works in practice, more concretely, how to specify a Cython dep in `pyproject.toml` as a build dependency in a way where the package is put in place for import mechanisms to find it but isn't actually built.

From my previous interactions with the pip developers they're pretty clueless and overwhelmed with finding any way to address this and just won't, so this might require some pushing and suggestions from the Cython project side.

### Code to reproduce the behaviour:

1. Make two packages that can be installed via setuptools `setup.py` or similar. One of them, let's call it package A, should have `.pxd` packages that end up being installed (via `MANIFEST.in` or such) into where the final module folder then is in site packages or alike.

2. Make the other package, package B, then depend on package A by 1. adding to its `setup.py` a mechanism that uses python's standard mechanisms, currently this would be importlib's `importlib.util.find_spec` I believe, to locate package A, then 2. use `cythonize`'s `include_path` option to specify that as a search path for `.pxd` files, then 3. `cimport` some `.pxd` from package A in package B somewhere. Please note you might run into #2886 and somehow need to work around that if your project also uses `.pyx`/`.pxd` for the `__init__.pyx` main module file.

3. You'll find that you'll need to add package A to package B's `pyproject.toml` as a build dependency or it won't be found with build isolation enabled. But that will also cause package A to be built even when the build may take hours (that's not an exaggeration for what I've run into on some devices). You'll probably also find out like me that there doesn't seem to be any mechanism to avoid this, e.g. some "this is a build dependency but don't build it itself" flag in `pyproject.toml`. You'll find that otherwise you'll need to tell end users to use `--no-build-isolation` which they won't understand why they need it, and also it might cause other issues down the line, like less reproducible and clean builds hiding other bugs, etc.

### Expected behaviour

There should be an obvious way how to specify a Cython project as a build dependency in `pyproject.toml` with some standardized flag that e.g. makes all `cythonize` calls a no-op by default or something, or something else that requires as little modifications to be widely supported by Cython projects as possible.

### OS

All OSes affected

### Python version

All Python versions affected if your pip is new enough to enforce build isolation

### Cython version

All Cython versions affected

**Comment by ell1e** [comment]
(Maybe I should note the kind of severity on this. Basically I sunk literal days into trying to make my project not a pain to install that I planned to open-source, a pretty big multiple thousands of lines graphical program of a few years, and I failed. Because I don't want to require pages of extra ways for a normal user to even try to get it running, I gave up on it for now, waiting for the Cython and pip ecosystem to somehow figure out a non-painful way to do this for both the dev and the user where I feel comfortable subjecting others to my code and the process of getting it to run.)

**Comment by scoder** [comment]
Thanks for the report. Let me try to understand what goes wrong here.

> that will also cause package A to be built even when the build may take hours

Doesn't pip install the other project as wheel? Or does this only apply to packages that do not provide wheels (for a given architecture)?

**Comment by ell1e** [comment]
Thanks for the response! As for wheels, yeah, this is my own stuff and I don't know how to provide wheels. :laughing:  It's also not on pypi. And wheels wouldn't solve it for my own installs of the dev versions to debug stuff, I think.

Also, pip hates cross compilation from last time I've tried it since it doesn't seem to understand when it's pointed at using a cross compiler and can't be made to, which at least from what I've gathered they also intend never to solve (which I think combined with build isolation is a big mistake but that's a whole other can of worms and less so connected to Cython), so that alone would get in the way of me ever reliably providing wheels ever.

But if all that means as a result people can't reasonably install it at all with hours of extra time, that's not really a way I can ever dare release this to others. Right now you basically need to specify `--no-build-isolation` and if you don't it breaks, but no regular user is going to understand that, so this is kind of untenable.

**Comment by scoder** [comment]
Wheels are relatively easy to build with cibuildwheel, e.g. on a Github Actions workflow. See [what Cython does](https://github.com/cython/cython/blob/master/.github/workflows/wheels.yml), for example. You can also cross-compile them for Arm64 and others. That would probably solve your problem, because installing wheels is quick.

As for a development workflow, maybe you can get away with not making your own packages (for which you have a local checkout) build dependencies of your project at all, but to just include their pxd directories in your PYTHONPATH, so that Cython can pick them up? You don't really have to build a package with pip while you are locally working on it.

**Comment by ell1e** [comment]
Thanks for the idea! I think the problem is that the `setup.py` would need to do that, right? So how would I do that in some friendly maintainable way when I just have the `requirements.txt` entry to pick up?

I'm not looking to find yet another barely maintainable hack (like the one to work around #2886 ) that makes it unbuildable for others.

**Comment by scoder** [comment]
Others will probably not have your other packages checked out to build together with those locally. If you provide binary wheels, your users will be fine. Probably you as well, as long as you're not making changes to the pxd files compared to what the binary wheels include. If you make change there, you'll have to point Cython to the local directory where you keep the updated versions.

**Comment by ell1e** [comment]
So is the answer really "sorry, the package just can't be built normally from source then"? :cry:

**Comment by ell1e** [comment]
I thought about this some more, and I think the main problem is that I don't understand why pip rebuilds something from scratch if it's already built and in the site packages. Doing so causes two big issues: 1. it risks for Cython in particular to rebuild with the wrong and not exact matching version causing crashes and corruption if there's e.g. `.pxd` importing going on but then with a binary incompatible variant, 2. it wastes lots of time even for non-Cython stuff, 3. it doesn't achieve anything of use if it was already just built. Why doesn't it generate a wheel locally (I think it actually does, usually) when that dependency is first installed into site packages, and then just reuse that wheel without rebuilding when inside the build isolation venv for the next thing depending on it? I don't understand that.

I'm probably missing something, but it's all really weird and cumbersome. But hacking around to search `.pxd` from a counter-intuitive third location that is neither `requirements.txt` or `pyproject.toml` manually sure doesn't sound like the correct way to deal with it.

**Comment by mattip** [comment]
Not really a cython issue, maybe we should continue this on the mailing list and not here.

pip is a package manager. It passes the build task on to a build backend, which in your case is setuptools/setup.py. When you tell pip to invoke the backend to build your package, it wants to do so in a reproducible way. That means creating an isolated environment to build in. (You can override this with `--no-build-isolation`). Now if your package has dependencies, and those dependencies are wheels, and your internet connection is OK, this isolation is not that painful: it will download and install a handful of wheels (which pip even caches locally to make it faster next time) and then pass the build task off to the build backend. But if you do not have wheels, pip, as a package manager, will look at the dependencies and build each one separately from source. You cannot really tell pip "hey, this is a dependency but it is not a package". If you want something like that, maybe you should not declare dependencies in your pyproject.toml. There are other choices:
- add them to your source tree as a submodule
- use a build system that knows about dependencies like bazel

So you have to choose: if you use pip to manage your dependencies, you have to buy into their model. I use "buy" intentionally: there is a cost to using the standard.

**Comment by ell1e** [comment]
> You cannot really tell pip "hey, this is a dependency but it is not a package".

But couldn't this be solved? All that would be needed is a flag that tells pip that it should still fetch this, place it somewhere, but just not build it. Or alternatively, to make it not rebuild it if it's already in the site packages but simply copy it over.

**Comment by ell1e** [comment]
Basically, when I started this project setuptool's setup_requires was the way to go and there was no build isolation, and everything worked fine even for `.pxd` imports without hours of rebuilds and without using wheels or anything. Then all these changes happened, and suddenly there is no way to do this anymore, and I just feel like it renders Cython way less practical for huge projects. With how this situation is today, I wouldn't have used Cython anymore, and I think that might be worth solving.

**Comment by mattip** [comment]
Where is the code to your project? Please open an issue there and ping me so we can continue the discussion somewhere else.

**Comment by ell1e** [comment]
It's currently not set to public because I saw no point, given I currently don't see a way how to make it installable in a way the average pip user understands :woman_shrugging: (which is why I made this ticket here).

Maybe I'm wrong, but I don't think this discussion is very dependent on the actual code. I think all projects that consist of 2+ larger Cython packages that take a while to build where one larger package does `.pxd` cimports from another larger one will run into this issue. Even with wheels this isn't fixable for building dev versions.

**Comment by scoder** [comment]
> Even with wheels this isn't fixable for building dev versions.

It sounds to me like the only issue that you have is local development. As soon as you release a public version with wheels, the problem goes away.

I don't currently see how we could help you here, so I'll close this ticket. If we find new ideas of what Cython specifically could do to improve the in-development situation of interrelated projects, it's fine to continue the discussion here. Or open a new ticket.

**Comment by ell1e** [comment]
Well, I was hoping you could contact the pip devs and figure out something that works for both sides. I think it's greatly harming the Cython ecosystem, I can't be the only one who finds this kind of untenable in daily work.

>  As soon as you release a public version with wheels, the problem goes away.

That really seems like window dressing, development work never stops. "Local development" being the "only" issue is a pretty massive issue. Also, small headcount projects simply may not have the resources for a CI to make wheels for all platforms it's used on.

**Comment by mattip** [comment]
> I think all projects that consist of ...

Do you have supporting examples of such projects? A counter example: SciPy [cimports numpy](https://github.com/scipy/scipy/blob/38410650cc0b0ffb058ec57b154a0bd688db8b8d/scipy/io/matlab/_mio_utils.pyx#L6), and successfully submits NumPy issues based on developer workflows when NumPy breaks their project, which is often.

> I can't be the only one who finds this pretty untenable for daily work

So far it seems so? Without code, it is very difficult to help you or even understand your situation. Feel free to invite me (mattip on github and codeberg) to your private repo so we can continue the conversation there, I can sign an NDA if needed.

> small headcount projects simply may not have the resources for a CI to make wheels ..

[cibuildwheel](https://cibuildwheel.readthedocs.io/en/stable/) goes a long way to making this work across many CI providers, and dovetails well with the [meson](https://mesonbuild.com/) build system plus ninja to build large cython projects like SciPy or Pandas in parallel. We recently converted NumPy from setup.py/setuptools to meson, as did Scipy, and are happy with the results: build time decreased by at least a factor of 5 if not more.

**Comment by GalaxySnail** [comment]
Could you provide an example of reproducing this issue by simply running `git clone` and `pip install`? AFAIK, [pip will try to use cached wheels by default](https://pip.pypa.io/en/latest/topics/caching/). If nothing is going wrong, pip shouldn't build the dependency twice.

**Comment by ell1e** [comment]
> AFAIK, [pip will try to use cached wheels by default](https://pip.pypa.io/en/latest/topics/caching/). If nothing is going wrong, pip shouldn't build the dependency twice.

Ah, yeah, that's an important piece of info, that would certainly explain why nobody else is complaining. For me somehow that doesn't work with my current setup.

There's certainly no NDA needed, I just need to make the time to clean up the copyright and license notices to set it public. And/or make a minimal example project to reproduce issue, since apparently it works for everyone else.

**Comment by ell1e** [comment]
> Installing directly from a source distribution will make pip build a wheel, regardless of whether there is a matching cached wheel.

I think this might be the problem. I'm always referring to the git repository since this package isn't on pypi, I assume from the quote that pip then just doesn't cache which matches what I'm seeing. Is there a way to fix that?

**Comment by ell1e** [comment]
I have as a consequence of this already decided a while ago that the Python ecosystem, with its combination of slowness and Cython becoming untenable with standard tooling, is better left behind. But it's sad, there's a lot of years of code and work that others could benefit from wasted. So I'm still hoping this will be resolved, but I can't say that I'm particularly positive. There seems to be a bit of apathy from all sides about this problem.

**Comment by ell1e** [comment]
For what it's worth:

> So far it seems so?

It seems many devs are generally are fed up with where pip is moving, at the very least, from those I talk to. I'm glad that I'm now heading elsewhere myself. I'm hoping some time the packaging situation of Python will improve, but in the last years it mostly seemed to only get worse and worse which is a shame.
